src/backend/dx11/shaders/copy.hlsl

```diff
@@ -1,15 +1,200 @@
+    data.y = min(data.y, 0x000000ff);
+    data.z = min(data.z, 0x000000ff);
+    data.w = min(data.w, 0x000000ff);
+
+    uint output = (data.x        |
+                  (data.y << 8)  |
+                  (data.z << 16) |
+                  (data.w << 24));
+
+    return output;
+}
+
+uint4 UintToUint4(uint data)
+{
+    return uint4((data & 0xff000000) >> 24, (data & 0xff0000) >> 16, (data & 0xff00) >> 8, data & 0xff);
+}
+
+uint2 UintToUint2(uint data)
+{
+    return uint2((data >> 16) & 0x0000ffff, data & 0x0000ffff);
+}
+
+uint Uint2ToUint(uint2 data)
+{
+    data.x = min(data.x, 0x0000ffff);
+    data.y = min(data.y, 0x0000ffff);
+
+    uint output = (data.x         |
+                  (data.y << 16));
+
+    return output;
+}
+
+// Buffers are always R32-aligned
+StructuredBuffer<uint> BufferCopySrc : register(t0);
+RWBuffer<uint>         BufferCopyDst: register(u0);
+
+// R32
+Texture2D<uint>    ImageCopySrcR32 : register(t0);
+RWTexture2D<uint>  ImageCopyDstR32 : register(u0);
+
+// TODO: correct, but slow
+[numthreads(1, 1, 1)]
+void cs_copy_buffer_image2d_r32(uint3 dispatch_thread_id : SV_DispatchThreadID) {
+    uint2 dst_idx = GetImageDst(dispatch_thread_id);
+    uint src_idx = GetBufferSrc(dispatch_thread_id);
+
+    ImageCopyDstR32[dst_idx] = BufferCopySrc[src_idx];
+}
+
+[numthreads(1, 1, 1)]
+void cs_copy_image2d_r32_buffer(uint3 dispatch_thread_id : SV_DispatchThreadID) {
+    uint dst_idx = GetBufferDst(dispatch_thread_id);
+    uint2 src_idx = GetImageSrc(dispatch_thread_id);
+
+    BufferCopyDst[dst_idx] = ImageCopySrcR32[src_idx];
+}
+
+// R16G16
+Texture2D<uint2>    ImageCopySrcR16G16 : register(t0);
+RWTexture2D<uint2>  ImageCopyDstR16G16 : register(u0);
 
 // TODO: correct, but slow
 [numthreads(1, 1, 1)]
-void cs_copy_buffer_image_2d(uint3 dispatch_thread_id : SV_DispatchThreadID) {
-    uint2 idx = ImageOffset + dispatch_thread_id.xy;
+void cs_copy_buffer_image2d_r16g16(uint3 dispatch_thread_id : SV_DispatchThreadID) {
+    uint2 dst_idx = GetImageDst(dispatch_thread_id);
+    uint src_idx = GetBufferSrc(dispatch_thread_id);
+
+    ImageCopyDstR16G16[dst_idx] = UintToUint2(BufferCopySrc[src_idx]);
+}
+
+[numthreads(1, 1, 1)]
+void cs_copy_image2d_r16g16_buffer(uint3 dispatch_thread_id : SV_DispatchThreadID) {
+    uint dst_idx = GetBufferDst(dispatch_thread_id);
+    uint2 src_idx = GetImageSrc(dispatch_thread_id);
+
+    BufferCopyDst[dst_idx] = Uint2ToUint(ImageCopySrcR16G16[src_idx].yx);
+}
+
+// R16
+Texture2D<uint>   ImageCopySrcR16 : register(t0);
+RWTexture2D<uint> ImageCopyDstR16 : register(u0);
+
+[numthreads(1, 1, 1)]
+void cs_copy_buffer_image2d_r16(uint3 dispatch_thread_id : SV_DispatchThreadID) {
+    uint src_idx = BufferImageCopies.BufferVars.x + dispatch_thread_id.x + dispatch_thread_id.y * BufferImageCopies.BufferVars.y / 2;
+
+    uint2 data = UintToUint2(BufferCopySrc[src_idx]);
+
+    ImageCopyDstR16[GetImageDst(uint3(2, 1, 0) * dispatch_thread_id + uint3(0, 0, 0))] = data.y;
```

Comment:
not sure I understand this. The image is `R16`, why do we need to write into 2 different texels?

Reply:
`R32_UINT` is the only format which has guaranteed UAV typed load support[0], so I was aiming for 1 buffer (source) access per thread, so in the case of `R16` one thread would have to do 2 writes. The dispatch call is also scaled depending on the format.

0: https://msdn.microsoft.com/en-us/library/windows/desktop/mt427455(v=vs.85).aspx